File: win/directx/sdk-9.0/Samples/VB.Net/Direct3D/BumpMapping/Emboss/Emboss.vb
'-----------------------------------------------------------------------------
' File: Emboss.vb
'
' Desc: Shows how to do a bumpmapping technique called embossing, in which a
'       heightmap is subtracted from itself, with slightly offset texture
'       coordinates for the second pass.
'
' Copyright (c) Microsoft Corporation. All rights reserved.
'-----------------------------------------------------------------------------
Imports System
Imports System.Windows.Forms
Imports Microsoft.DirectX
Imports Microsoft.DirectX.Direct3D


Namespace BumpWavesSample
    _



    '/ <summary>
    '/ Global Emboss structure
    '/ </summary>
    Public Structure EmbossVertex
        Public p As Vector3
        Public n As Vector3
        Public tu, tv As Single
        Public tu2, tv2 As Single

        Public Const Format As VertexFormats = VertexFormats.PositionNormal Or VertexFormats.Texture2
    End Structure 'EmbossVertex
    _


    '/ <summary>
    '/ Application class. The base class (GraphicsSample) provides the generic 
    '/ functionality needed in all Direct3D samples. MyGraphicsSample adds 
    '/ functionality specific to this sample program.
    '/ </summary>
    Public Class MyGraphicsSample
        Inherits GraphicsSample



        Private drawingFont As GraphicsFont = Nothing ' Font for drawing text
        ' Scene
        Private renderObject As GraphicsMesh = Nothing ' Object to render
        Private isShowingEmbossMethod As Boolean = True ' Whether to do the embossing
        Private embossTexture As Texture ' The emboss texture
        Private bumpLightPos As Vector3 ' Light position
        Private tangents As Vector3() = Nothing ' Array of vertex tangents
        Private binormals As Vector3() = Nothing ' Array of vertex binormals
        Private worldMatrix As Matrix ' The world matrix

        ' Menu items for this sample
        Private mnuOptions As System.Windows.Forms.MenuItem
        Private mnuEmboss As System.Windows.Forms.MenuItem


        '/ <summary>
        '/ Application constructor. Sets attributes for the app.
        '/ </summary>
        Public Sub New()
            ' Set the window text
            Me.Text = "Emboss: BumpMapping Technique"
            Try
                ' Load the icon from our resources
                Dim resources As New System.Resources.ResourceManager(Me.GetType())
                Me.Icon = CType(resources.GetObject("$this.Icon"), System.Drawing.Icon)
            Catch
            End Try
            ' It's no big deal if we can't load our icons

            drawingFont = New GraphicsFont("Arial", System.Drawing.FontStyle.Bold)
            enumerationSettings.AppUsesDepthBuffer = True


            ' Add our new menu options
            mnuOptions = New System.Windows.Forms.MenuItem("&Options")
            mnuEmboss = New System.Windows.Forms.MenuItem("Toggle &emboss mode")
            ' Add to the main menu screen
            mnuMain.MenuItems.Add(Me.mnuOptions)
            mnuOptions.MenuItems.Add(mnuEmboss)
            mnuEmboss.Shortcut = System.Windows.Forms.Shortcut.CtrlE
            mnuEmboss.ShowShortcut = True
            AddHandler mnuEmboss.Click, AddressOf Me.EmbossModeChanged
        End Sub 'New





        '/ <summary>
        '/ To find a tangent that heads in the direction of +tv(texcoords), find 
        '/ the components of both vectors on the tangent surface, and add a 
        '/ linear combination of the two projections that head in the +tv direction
        '/ </summary>
        Function ComputeTangentVector(ByVal vertexA As EmbossVertex, ByVal vertexB As EmbossVertex, ByVal vertexC As EmbossVertex) As Vector3
            Dim vAB As Vector3 = Vector3.Subtract(vertexB.p, vertexA.p)
            Dim vAC As Vector3 = Vector3.Subtract(vertexC.p, vertexA.p)
            Dim n As Vector3 = vertexA.n

            ' Components of vectors to neghboring vertices that are orthogonal to the
            ' vertex normal
            Dim vProjAB As Vector3 = Vector3.Subtract(vAB, Vector3.Multiply(n, Vector3.Dot(n, vAB)))
            Dim vProjAC As Vector3 = Vector3.Subtract(vAC, Vector3.Multiply(n, Vector3.Dot(n, vAC)))

            ' tu and tv texture coordinate differences
            Dim duAB As Single = vertexB.tu - vertexA.tu
            Dim duAC As Single = vertexC.tu - vertexA.tu
            Dim dvAB As Single = vertexB.tv - vertexA.tv
            Dim dvAC As Single = vertexC.tv - vertexA.tv

            If duAC * dvAB > duAB * dvAC Then
                duAC = -duAC
                duAB = -duAB
            End If

            Dim vTangent As Vector3 = Vector3.Subtract(Vector3.Multiply(vProjAB, duAC), Vector3.Multiply(vProjAC, duAB))
            vTangent.Normalize()
            Return vTangent
        End Function 'ComputeTangentVector





        '/ <summary>
        '/ Compute the tangents and Binormals
        '/ </summary>
        Sub ComputeTangentsAndBinormals()
            Dim vertices As EmbossVertex() = Nothing
            Dim indices As Short() = Nothing

            Dim numVertices As Integer = 0
            Dim numIndices As Integer = 0

            ' Gain access to the object's vertex and index buffers
            Dim vertexBuffer As VertexBuffer = renderObject.SystemVertexBuffer
            numVertices = renderObject.SysMemMesh.NumberVertices
            vertices = CType(vertexBuffer.Lock(0, GetType(EmbossVertex), 0, numVertices), EmbossVertex())

            Dim indexBuffer As IndexBuffer = renderObject.SystemIndexBuffer
            numIndices = renderObject.SysMemMesh.NumberFaces * 3
            indices = CType(indexBuffer.Lock(0, GetType(Short), 0, numIndices), Short())

            ' Allocate space for the vertices' tangents and binormals
            tangents = New Vector3(numVertices) {}
            binormals = New Vector3(numVertices) {}

            ' Generate the vertices' tangents and binormals
            Dim i As Integer
            For i = 0 To numIndices - 3 Step 3
                Dim a As Short = indices((i + 0))
                Dim b As Short = indices((i + 1))
                Dim c As Short = indices((i + 2))

                ' To find a tangent that heads in the direction of +tv(texcoords),
                ' find the components of both vectors on the tangent surface ,
                ' and add a linear combination of the two projections that head in the +tv direction
                tangents(a) = Vector3.Add(tangents(a), ComputeTangentVector(vertices(a), vertices(b), vertices(c)))
                tangents(b) = Vector3.Add(tangents(b), ComputeTangentVector(vertices(b), vertices(a), vertices(c)))
                tangents(c) = Vector3.Add(tangents(c), ComputeTangentVector(vertices(c), vertices(a), vertices(b)))
            Next i

            For i = 0 To numVertices - 1
                ' Normalize the tangents
                tangents(i).Normalize()

                ' Compute the binormals
                binormals(i) = Vector3.Cross(vertices(i).n, tangents(i))
            Next i

            ' Unlock the buffers
            vertexBuffer.Unlock()
            indexBuffer.Unlock()
        End Sub 'ComputeTangentsAndBinormals





        '/ <summary>
        '/ Performs a calculation on each of the vertices' normals to determine
        '/ what the texture coordinates should be for the environment map (in this 
        '/ case the bump map).
        '/ </summary>
        Sub ApplyEnvironmentMap()
            Dim vertices As EmbossVertex() = Nothing
            Dim numVertices As Integer = 0

            ' Gain access to the object's vertex and index buffers
            Dim vertexBuffer As VertexBuffer = renderObject.LocalVertexBuffer
            numVertices = renderObject.LocalMesh.NumberVertices
            vertices = CType(vertexBuffer.Lock(0, GetType(EmbossVertex), 0, numVertices), EmbossVertex())

            ' Get an inverse world matrix
            Dim invertWorld As Matrix = Matrix.Invert(worldMatrix)

            ' Get the current light position in object space
            Dim transformed As Vector4 = Vector3.Transform(device.Lights(0).Position, invertWorld)
            bumpLightPos.X = transformed.X
            bumpLightPos.Y = transformed.Y
            bumpLightPos.Z = transformed.Z

            ' Dimensions of texture needed for shifting tex coords
            Dim surfacedesc As SurfaceDescription = embossTexture.GetLevelDescription(0)

            ' Loop through the vertices, transforming each one and calculating
            ' the correct texture coordinates.
            Dim i As Short
            For i = 0 To numVertices - 1
                ' Find light vector in tangent space
                Dim vDir As Vector3 = Vector3.Subtract(bumpLightPos, vertices(i).p)
                Dim vLightToVertex As Vector3 = Vector3.Normalize(vDir)

                ' Create rotation matrix (rotate into tangent space)
                Dim r As Single = Vector3.Dot(vLightToVertex, vertices(i).n)

                If r < 0.0F Then
                    ' Don't shift coordinates when light below surface
                    vertices(i).tu2 = vertices(i).tu
                    vertices(i).tv2 = vertices(i).tv
                Else
                    ' Shift coordinates for the emboss effect
                    Dim vEmbossShift As Vector2 = Vector2.Empty
                    vEmbossShift.X = Vector3.Dot(vLightToVertex, tangents(i))
                    vEmbossShift.Y = Vector3.Dot(vLightToVertex, binormals(i))
                    vEmbossShift.Normalize()
                    vertices(i).tu2 = vertices(i).tu + vEmbossShift.X / device.PresentationParameters.BackBufferWidth
                    vertices(i).tv2 = vertices(i).tv - vEmbossShift.Y / device.PresentationParameters.BackBufferHeight
                End If
            Next i

            vertexBuffer.Unlock()
        End Sub 'ApplyEnvironmentMap




        '/ <summary>
        '/ Called once per frame, the call is the entry point for animating the scene.
        '/ </summary>
        Protected Overrides Sub FrameMove()
            ' Rotate the object
            worldMatrix = Matrix.RotationY(appTime)
            Device.Transform.World = worldMatrix
        End Sub 'FrameMove





        '/ <summary>
        '/ Called once per frame, the call is the entry point for 3d rendering. This 
        '/ function sets up render states, clears the viewport, and renders the scene.
        '/ </summary>
        Protected Overrides Sub Render()
            ' Clear the viewport
            Device.Clear(ClearFlags.Target Or ClearFlags.ZBuffer, System.Drawing.Color.Black, 1.0F, 0)

            Device.BeginScene()


            ' Stage 0 is the base texture, with the height map in the alpha channel
            Device.SetTexture(0, embossTexture)
            Device.TextureState(0).TextureCoordinateIndex = 0
            Device.TextureState(0).ColorOperation = TextureOperation.Modulate
            Device.TextureState(0).ColorArgument1 = TextureArgument.TextureColor
            Device.TextureState(0).ColorArgument2 = TextureArgument.Diffuse
            Device.TextureState(0).AlphaOperation = TextureOperation.SelectArg1
            Device.TextureState(0).AlphaArgument1 = TextureArgument.TextureColor

            If isShowingEmbossMethod Then
                ' Stage 1 passes through the RGB channels (SelectArg2 = Current), and 
                ' does a signed add with the inverted alpha channel. The texture coords
                ' associated with Stage 1 are the shifted ones, so the result is:
                '    (height - shifted_height) * tex.RGB * diffuse.RGB
                Device.SetTexture(1, embossTexture)
                Device.TextureState(1).TextureCoordinateIndex = 1
                Device.TextureState(1).ColorOperation = TextureOperation.SelectArg2
                Device.TextureState(1).ColorArgument1 = TextureArgument.TextureColor
                Device.TextureState(1).ColorArgument2 = TextureArgument.Current
                Device.TextureState(1).AlphaOperation = TextureOperation.AddSigned
                Device.TextureState(1).AlphaArgument1 = TextureArgument.TextureColor Or TextureArgument.Complement
                Device.TextureState(1).AlphaArgument2 = TextureArgument.Current

                ' Set up the alpha blender to multiply the alpha channel (monochrome emboss)
                ' with the src color (lighted texture)
                Device.RenderState.AlphaBlendEnable = True
                Device.RenderState.SourceBlend = Blend.SourceAlpha
                Device.RenderState.DestinationBlend = Blend.Zero
            End If

            ' Render the object
            renderObject.Render(Device)

            ' Restore render states
            Device.TextureState(1).ColorOperation = TextureOperation.Disable
            Device.TextureState(1).AlphaOperation = TextureOperation.Disable
            Device.RenderState.AlphaBlendEnable = False

            ' Output statistics
            drawingFont.DrawText(2, 1, System.Drawing.Color.Yellow, frameStats)
            drawingFont.DrawText(2, 20, System.Drawing.Color.Yellow, deviceStats)
            drawingFont.DrawText(2, 40, System.Drawing.Color.Yellow, "Move the light with the mouse")
            drawingFont.DrawText(2, 60, System.Drawing.Color.Yellow, "Emboss-mode:")
            drawingFont.DrawText(130, 60, System.Drawing.Color.White, IIf(isShowingEmbossMethod, "ON", "OFF"))

            Device.EndScene()
        End Sub 'Render





        '/ <summary>
        '/ Initialize scene objects.
        '/ </summary>
        Protected Overrides Sub InitializeDeviceObjects()
            drawingFont.InitializeDeviceObjects(Device)
            Try
                ' Load texture map. Note that this is a special texture, which has a
                ' height field stored in the alpha channel
                embossTexture = GraphicsUtility.CreateTexture(Device, "emboss1.dds", Format.A8R8G8B8)

                ' Load geometry
                renderObject = New GraphicsMesh()
                renderObject.Create(Device, "tiger.x")

                ' Set attributes for geometry
                renderObject.SetFVF(Device, EmbossVertex.Format)
                renderObject.UseMeshMaterials = False

                ' Compute the object's tangents and binormals, whaich are needed for the 
                ' emboss-tecnhique's texture-coordinate shifting calculations
                ComputeTangentsAndBinormals()
            Catch
                Dim e As MediaNotFoundException = New MediaNotFoundException()
                HandleSampleException(e, ApplicationMessage.ApplicationMustExit)
                Throw e
            End Try
        End Sub 'InitializeDeviceObjects





        '/ <summary>
        '/ 
        '/ </summary>
        Protected Overrides Sub RestoreDeviceObjects(ByVal sender As System.Object, ByVal e As System.EventArgs)
            renderObject.RestoreDeviceObjects(Device, Nothing)
            ' Set up the textures
            Device.TextureState(0).ColorArgument1 = TextureArgument.TextureColor
            Device.TextureState(0).ColorArgument2 = TextureArgument.Diffuse
            Device.TextureState(0).ColorOperation = TextureOperation.Modulate
            Device.SamplerState(0).MinFilter = TextureFilter.Linear
            Device.SamplerState(0).MagFilter = TextureFilter.Linear

            ' Set the transform matrices
            Dim vEyePt As New Vector3(0.0F, 0.0F, 3.5F)
            Dim vLookatPt As New Vector3(0.0F, 0.0F, 0.0F)
            Dim vUpVec As New Vector3(0.0F, 1.0F, 0.0F)

            Dim matView As Matrix = Matrix.LookAtLH(vEyePt, vLookatPt, vUpVec)
            Dim aspect As Single = Device.PresentationParameters.BackBufferWidth / Device.PresentationParameters.BackBufferHeight

            Dim matProj As Matrix = Matrix.PerspectiveFovLH(CSng(Math.PI) / 4, aspect, 1.0F, 1000.0F)
            Device.Transform.View = matView
            Device.Transform.Projection = matProj

            ' Setup a material
            Device.Material = GraphicsUtility.InitMaterial(System.Drawing.Color.White)

            ' Setup a light
            GraphicsUtility.InitLight(Device.Lights(0), LightType.Point, 5.0F, 5.0F, -20.0F)
            Device.Lights(0).Attenuation0 = 1.0F
            Device.Lights(0).Commit()
            Device.Lights(0).Enabled = True

            ' Set miscellaneous render states
            Device.RenderState.ZBufferEnable = True
            Device.RenderState.Lighting = True
            Device.RenderState.Ambient = System.Drawing.Color.FromArgb(&H444444)

            ApplyEnvironmentMap()
        End Sub 'RestoreDeviceObjects





        '/ <summary>
        '/ Called during device initialization, this code checks the device for some 
        '/ minimum set of capabilities
        '/ </summary>
        Protected Overrides Function ConfirmDevice(ByVal caps As Microsoft.DirectX.Direct3D.Caps, ByVal vertexProcessingType As VertexProcessingType, ByVal adapterFmt As Microsoft.DirectX.Direct3D.Format, ByVal backBufferFmt As Microsoft.DirectX.Direct3D.Format) As Boolean
            ' This sample uses the ADDSIGNED texture blending mode
            If Not caps.TextureOperationCaps.SupportsAddSigned Then
                Return False
            End If
            If caps.MaxTextureBlendStages < 2 Then
                Return False
            End If
            Return True
        End Function 'ConfirmDevice







        '/ <summary>
        '/ Move the light with the mouse
        '/ </summary>
        Protected Overrides Sub OnMouseMove(ByVal e As System.Windows.Forms.MouseEventArgs)
            Dim w As Single = Device.PresentationParameters.BackBufferWidth
            Dim h As Single = Device.PresentationParameters.BackBufferHeight
            Device.Lights(0).XPosition = 200.0F * (0.5F - e.X / w)
            Device.Lights(0).YPosition = 200.0F * (0.5F - e.Y / h)
            Device.Lights(0).ZPosition = 100.0F
            Device.Lights(0).Commit()
            ApplyEnvironmentMap()

            ' Let the control handle the mouse now
            MyBase.OnMouseMove(e)
        End Sub 'OnMouseMove




        '/ <summary>
        '/ Fired when the emboss mode has changed
        '/ </summary>
        Private Sub EmbossModeChanged(ByVal sender As Object, ByVal e As EventArgs)
            isShowingEmbossMethod = Not isShowingEmbossMethod
        End Sub 'EmbossModeChanged





        '/ <summary>
        '/ The main entry point for the application.
        '/ </summary>
        Shared Sub Main()
            Dim d3dApp As New MyGraphicsSample()
            Try
                If d3dApp.CreateGraphicsSample() Then
                    d3dApp.Run()
                End If
            Finally
                d3dApp.Dispose()
            End Try
        End Sub 'Main 
    End Class 'MyGraphicsSample
End Namespace 'BumpWavesSample

File: win/directx/sdk-9.0/Samples/C#/Direct3D/BumpMapping/Emboss/Emboss.cs
//-----------------------------------------------------------------------------
// File: Emboss.cs
//
// Desc: Shows how to do a bumpmapping technique called embossing, in which a
//       heightmap is subtracted from itself, with slightly offset texture
//       coordinates for the second pass.
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
using System;
using System.Windows.Forms;
using Microsoft.DirectX;
using Microsoft.DirectX.Direct3D;




namespace BumpWavesSample
{
    /// <summary>
    /// Global Emboss structure
    /// </summary>
    public struct EmbossVertex
    {
        public Vector3 p;
        public Vector3 n;
        public float tu, tv;
        public float tu2, tv2;

        public const VertexFormats Format = VertexFormats.PositionNormal | VertexFormats.Texture2;
    }

	


	/// <summary>
	/// Application class. The base class (GraphicsSample) provides the generic 
	/// functionality needed in all Direct3D samples. MyGraphicsSample adds 
	/// functionality specific to this sample program.
	/// </summary>
	public class MyGraphicsSample : GraphicsSample
	{
		private GraphicsFont drawingFont = null; // Font for drawing text
		// Scene
        GraphicsMesh renderObject = null; // Object to render
        bool isShowingEmbossMethod = true; // Whether to do the embossing

        Texture embossTexture; // The emboss texture
        Vector3 bumpLightPos; // Light position
        Vector3[] tangents = null; // Array of vertex tangents
        Vector3[] binormals = null; // Array of vertex binormals
        Matrix worldMatrix; // The world matrix

		
        // Menu items for this sample
        System.Windows.Forms.MenuItem mnuOptions;
        System.Windows.Forms.MenuItem mnuEmboss;


        
        
        /// <summary>
		/// Application constructor. Sets attributes for the app.
		/// </summary>
		public MyGraphicsSample()
		{
			// Set the window text
			this.Text = "Emboss: BumpMapping Technique";
            try
            {
                // Load the icon from our resources
                System.Resources.ResourceManager resources = new System.Resources.ResourceManager(this.GetType());
                this.Icon = ((System.Drawing.Icon)(resources.GetObject("$this.Icon")));
            }
            catch
            {
                // It's no big deal if we can't load our icons
            }

			drawingFont = new GraphicsFont("Arial", System.Drawing.FontStyle.Bold);
			enumerationSettings.AppUsesDepthBuffer = true;


            // Add our new menu options
            mnuOptions = new System.Windows.Forms.MenuItem("&Options");
            mnuEmboss = new System.Windows.Forms.MenuItem("Toggle &emboss mode");
            // Add to the main menu screen
            mnuMain.MenuItems.Add(this.mnuOptions);
            mnuOptions.MenuItems.Add(mnuEmboss);
            mnuEmboss.Shortcut = System.Windows.Forms.Shortcut.CtrlE;
            mnuEmboss.ShowShortcut = true;
            mnuEmboss.Click += new System.EventHandler(this.EmbossModeChanged);
        }




        /// <summary>
        /// To find a tangent that heads in the direction of +tv(texcoords), find 
        /// the components of both vectors on the tangent surface, and add a 
        /// linear combination of the two projections that head in the +tv direction
        /// </summary>
        Vector3 ComputeTangentVector(EmbossVertex vertexA, EmbossVertex vertexB,  EmbossVertex vertexC)
        {
            Vector3 vAB = vertexB.p - vertexA.p;
            Vector3 vAC = vertexC.p - vertexA.p;
            Vector3 n = vertexA.n;

            // Components of vectors to neghboring vertices that are orthogonal to the
            // vertex normal
            Vector3 vProjAB = vAB - (Vector3.Dot(n, vAB) * n);
            Vector3 vProjAC = vAC - (Vector3.Dot(n, vAC) * n);

            // tu and tv texture coordinate differences
            float duAB = vertexB.tu - vertexA.tu;
            float duAC = vertexC.tu - vertexA.tu;
            float dvAB = vertexB.tv - vertexA.tv;
            float dvAC = vertexC.tv - vertexA.tv;

            if (duAC*dvAB > duAB*dvAC)
            {
                duAC = -duAC;
                duAB = -duAB;
            }
    
            Vector3 vTangent = duAC*vProjAB - duAB*vProjAC;
            vTangent.Normalize();
            return vTangent;
        }




        /// <summary>
        /// Compute the tangents and Binormals
        /// </summary>
        void ComputeTangentsAndBinormals()
        {
            EmbossVertex[] vertices = null;
            short[] indices = null;

            int numVertices = 0;
            int numIndices = 0;

            // Gain access to the object's vertex and index buffers
            VertexBuffer vertexBuffer = renderObject.SystemVertexBuffer;
            numVertices = renderObject.SystemMesh.NumberVertices;
            vertices = (EmbossVertex[])vertexBuffer.Lock(0, typeof(EmbossVertex), 0, numVertices);

            IndexBuffer indexBuffer = renderObject.SystemIndexBuffer;
            numIndices = renderObject.SystemMesh.NumberFaces * 3;
            indices = (short[])indexBuffer.Lock(0, typeof(short), 0, numIndices);

            // Allocate space for the vertices' tangents and binormals
            tangents = new Vector3[numVertices];
            binormals = new Vector3[numVertices];

            // Generate the vertices' tangents and binormals
            for (int i = 0; i < numIndices; i += 3)
            {
                short a = indices[i+0];
                short b = indices[i+1];
                short c = indices[i+2];

                // To find a tangent that heads in the direction of +tv(texcoords),
                // find the components of both vectors on the tangent surface ,
                // and add a linear combination of the two projections that head in the +tv direction
                tangents[a] += ComputeTangentVector(vertices[a], vertices[b], vertices[c]);
                tangents[b] += ComputeTangentVector(vertices[b], vertices[a], vertices[c]);
                tangents[c] += ComputeTangentVector(vertices[c], vertices[a], vertices[b]);
            }

            for (int i = 0; i < numVertices; i++)
            {
                // Normalize the tangents
                tangents[i].Normalize();

                // Compute the binormals
                binormals[i] = Vector3.Cross(vertices[i].n, tangents[i]);
            }        

            // Unlock the buffers
            vertexBuffer.Unlock();
            indexBuffer.Unlock();
        }




        /// <summary>
        /// Performs a calculation on each of the vertices' normals to determine
        /// what the texture coordinates should be for the environment map (in this 
        /// case the bump map).
        /// </summary>
        void ApplyEnvironmentMap()
        {
            EmbossVertex[] vertices = null;
            int numVertices = 0;

            // Gain access to the object's vertex and index buffers
            VertexBuffer vertexBuffer = renderObject.LocalVertexBuffer;
            numVertices = renderObject.LocalMesh.NumberVertices;
            vertices = (EmbossVertex[])vertexBuffer.Lock(0, typeof(EmbossVertex), 0, numVertices);

            // Get an inverse world matrix
            Matrix invertWorld = Matrix.Invert(worldMatrix);

            // Get the current light position in object space
            Vector4 transformed = Vector3.Transform(device.Lights[0].Position, invertWorld);
            bumpLightPos.X = transformed.X;
            bumpLightPos.Y = transformed.Y;
            bumpLightPos.Z = transformed.Z;

            // Dimensions of texture needed for shifting tex coords
            SurfaceDescription surfacedesc = embossTexture.GetLevelDescription(0);
    
            // Loop through the vertices, transforming each one and calculating
            // the correct texture coordinates.
            for (short i = 0; i < numVertices; i++)
            {
                // Find light vector in tangent space
                Vector3 vDir = bumpLightPos - vertices[i].p;
                Vector3 vLightToVertex = Vector3.Normalize(vDir);
        
                // Create rotation matrix (rotate into tangent space)
                float r = Vector3.Dot(vLightToVertex, vertices[i].n);

                if (r < 0.0f) 
                {
                    // Don't shift coordinates when light below surface
                    vertices[i].tu2 = vertices[i].tu;
                    vertices[i].tv2 = vertices[i].tv;
                }
                else
                {
                    // Shift coordinates for the emboss effect
                    Vector2 vEmbossShift = Vector2.Empty;
                    vEmbossShift.X = Vector3.Dot(vLightToVertex, tangents[i]);
                    vEmbossShift.Y = Vector3.Dot(vLightToVertex, binormals[i]);
                    vEmbossShift.Normalize();
                    vertices[i].tu2 = vertices[i].tu + vEmbossShift.X / device.PresentationParameters.BackBufferWidth;
                    vertices[i].tv2 = vertices[i].tv - vEmbossShift.Y / device.PresentationParameters.BackBufferHeight;
                }
            }

            vertexBuffer.Unlock();
        }




		/// <summary>
		/// Called once per frame, the call is the entry point for animating the scene.
		/// </summary>
		protected override void FrameMove()
		{
            // Rotate the object
            worldMatrix = Matrix.RotationY(appTime);
            device.Transform.World = worldMatrix;
        }




		/// <summary>
		/// Called once per frame, the call is the entry point for 3d rendering. This 
		/// function sets up render states, clears the viewport, and renders the scene.
		/// </summary>
		protected override void Render()
		{
			// Clear the viewport
			device.Clear(ClearFlags.Target | ClearFlags.ZBuffer, System.Drawing.Color.Black, 1.0f, 0);

			device.BeginScene();


            // Stage 0 is the base texture, with the height map in the alpha channel
            device.SetTexture(0, embossTexture);
            device.TextureState[0].TextureCoordinateIndex = 0;
            device.TextureState[0].ColorOperation = TextureOperation.Modulate;
            device.TextureState[0].ColorArgument1 = TextureArgument.TextureColor;
            device.TextureState[0].ColorArgument2 = TextureArgument.Diffuse;
            device.TextureState[0].AlphaOperation = TextureOperation.SelectArg1;
            device.TextureState[0].AlphaArgument1 = TextureArgument.TextureColor;

            if (isShowingEmbossMethod)
            {
                // Stage 1 passes through the RGB channels (SelectArg2 = Current), and 
                // does a signed add with the inverted alpha channel. The texture coords
                // associated with Stage 1 are the shifted ones, so the result is:
                //    (height - shifted_height) * tex.RGB * diffuse.RGB
                device.SetTexture(1, embossTexture);
                device.TextureState[1].TextureCoordinateIndex = 1;
                device.TextureState[1].ColorOperation = TextureOperation.SelectArg2;
                device.TextureState[1].ColorArgument1 = TextureArgument.TextureColor;
                device.TextureState[1].ColorArgument2 = TextureArgument.Current;
                device.TextureState[1].AlphaOperation = TextureOperation.AddSigned;
                device.TextureState[1].AlphaArgument1 = TextureArgument.TextureColor | TextureArgument.Complement;
                device.TextureState[1].AlphaArgument2 = TextureArgument.Current;

                // Set up the alpha blender to multiply the alpha channel (monochrome emboss)
                // with the src color (lighted texture)
                device.RenderState.AlphaBlendEnable = true;
                device.RenderState.SourceBlend = Blend.SourceAlpha;
                device.RenderState.DestinationBlend = Blend.Zero;
            }

            // Render the object
            renderObject.Render(device);
    
            // Restore render states
            device.TextureState[1].ColorOperation = TextureOperation.Disable;
            device.TextureState[1].AlphaOperation = TextureOperation.Disable;
            device.RenderState.AlphaBlendEnable = false;

			// Output statistics
			drawingFont.DrawText(2,  1, System.Drawing.Color.Yellow, frameStats);
			drawingFont.DrawText(2, 20, System.Drawing.Color.Yellow, deviceStats);
            drawingFont.DrawText(2, 40, System.Drawing.Color.Yellow, "Move the light with the mouse");
            drawingFont.DrawText(2, 60, System.Drawing.Color.Yellow, "Emboss-mode:");
            drawingFont.DrawText(130, 60, System.Drawing.Color.White, isShowingEmbossMethod ? "ON" : "OFF");

			device.EndScene();
		}




		/// <summary>
        /// The device has been created.  Resources that are not lost on
        /// Reset() can be created here -- resources in Pool.Managed,
        /// Pool.Scratch, or Pool.SystemMemory.  Image surfaces created via
        /// CreateImageSurface are never lost and can be created here.  Vertex
        /// shaders and pixel shaders can also be created here as they are not
        /// lost on Reset().
		/// </summary>
		protected override void InitializeDeviceObjects()
		{
			drawingFont.InitializeDeviceObjects(device);
			try
			{
                // Load texture map. Note that this is a special texture, which has a
                // height field stored in the alpha channel
                embossTexture = GraphicsUtility.CreateTexture(device, "emboss1.dds", Format.A8R8G8B8);

                // Load geometry
                renderObject = new GraphicsMesh();
                renderObject.Create(device, "tiger.x");

                // Set attributes for geometry
                renderObject.SetVertexFormat(device, EmbossVertex.Format);
                renderObject.IsUsingMaterials = false;

                // Compute the object's tangents and binormals, whaich are needed for the 
                // emboss-tecnhique's texture-coordinate shifting calculations
                ComputeTangentsAndBinormals();
            }
			catch 
			{
				SampleException e = new MediaNotFoundException();
				HandleSampleException(e, ApplicationMessage.ApplicationMustExit);
				throw e;
			}
		}




		/// <summary>
        /// The device exists, but may have just been Reset().  Resources in
        /// Pool.Default and any other device state that persists during
        /// rendering should be set here.  Render states, matrices, textures,
        /// etc., that don't change during rendering can be set once here to
        /// avoid redundant state setting during Render() or FrameMove().
		/// </summary>
		protected override void RestoreDeviceObjects(System.Object sender, System.EventArgs e)
		{
            renderObject.RestoreDeviceObjects(device, null);
            // Set up the textures
            device.TextureState[0].ColorArgument1 = TextureArgument.TextureColor;
            device.TextureState[0].ColorArgument2 = TextureArgument.Diffuse;
            device.TextureState[0].ColorOperation = TextureOperation.Modulate;
            device.SamplerState[0].MinFilter = TextureFilter.Linear;
            device.SamplerState[0].MagFilter = TextureFilter.Linear;

			// Set the transform matrices
			Vector3 vEyePt = new Vector3(0.0f, 0.0f, 3.5f);
			Vector3 vLookatPt = new Vector3(0.0f, 0.0f, 0.0f);
			Vector3 vUpVec = new Vector3(0.0f, 1.0f, 0.0f);

            Matrix matView = Matrix.LookAtLH(vEyePt, vLookatPt, vUpVec);
            float aspect = device.PresentationParameters.BackBufferWidth / device.PresentationParameters.BackBufferHeight;

            Matrix matProj = Matrix.PerspectiveFovLH((float)Math.PI / 4, aspect, 1.0f, 1000.0f);
			device.Transform.View = matView;
			device.Transform.Projection = matProj;

            // Setup a material
            device.Material = GraphicsUtility.InitMaterial(System.Drawing.Color.White);

            // Setup a light
            GraphicsUtility.InitLight(device.Lights[0], LightType.Point, 5.0f, 5.0f, -20.0f);
            device.Lights[0].Attenuation0 = 1.0f;
            device.Lights[0].Commit();
            device.Lights[0].Enabled = true;

            // Set miscellaneous render states
            device.RenderState.ZBufferEnable = true;
            device.RenderState.Lighting = true;
            device.RenderState.Ambient = System.Drawing.Color.FromArgb(0x00444444);

            ApplyEnvironmentMap();
		}




		/// <summary>
		/// Called during device initialization, this code checks the device for some 
		/// minimum set of capabilities
		/// </summary>
		protected override bool ConfirmDevice(Caps caps, VertexProcessingType vertexProcessingType, 
			Format adapterFormat, Format backBufferFormat)
		{
            // This sample uses the ADDSIGNED texture blending mode
            if (!caps.TextureOperationCaps.SupportsAddSigned)
				return false;

            if (caps.MaxTextureBlendStages < 2)
                return false;

            return true;
		}




        /// <summary>
        /// Move the light with the mouse
        /// </summary>
        protected override void OnMouseMove(System.Windows.Forms.MouseEventArgs e)
        {
            float w = device.PresentationParameters.BackBufferWidth;
            float h = device.PresentationParameters.BackBufferHeight;
            device.Lights[0].XPosition = 200.0f * (0.5f - e.X / w);
            device.Lights[0].YPosition = 200.0f * (0.5f - e.Y / h);
            device.Lights[0].ZPosition = 100.0f;
            device.Lights[0].Commit();
            ApplyEnvironmentMap();

            // Let the control handle the mouse now
            base.OnMouseMove(e);
        }




        /// <summary>
        /// Fired when the emboss mode has changed
        /// </summary>
        private void EmbossModeChanged(object sender, EventArgs e)
        {
            isShowingEmbossMethod = !isShowingEmbossMethod;
        }



		
        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        static void Main() 
        {
            using (MyGraphicsSample d3dApp = new MyGraphicsSample())
            {                                 
                if (d3dApp.CreateGraphicsSample())
                    d3dApp.Run();
            }
        }
	}
}
